src/checker.rs

```diff
@@ -15,62 +21,162 @@ pub enum FailureReason {
     DnsError(String),
     Error(String),
 }
+
 /// Fetches the response from a URL.
 ///
 /// First attempts to fetch just the head, and if not supported falls back to fetching the entire body.
-async fn do_request(client: &reqwest::Client, url: &str) -> Result<Response, reqwest::Error> {
-    let response = client.head(url).send().await?;
-    match response.status() {
-        StatusCode::METHOD_NOT_ALLOWED => client.get(url).send().await,
-        _ => Ok(response),
+async fn do_request(
+    client: &reqwest::Client,
+    url: &str,
+) -> Result<(RequestType, Response), (RequestType, reqwest::Error)> {
```

Comment:
I don't totally love this

Reply:
I would change this to
```suggestion
) -> (RequestType, Result<Response, reqwest::Error>) {
```
You can't use `?` in the function anymore, but it's not too bad:
```rust
    let response = match client.head(url).send().await {
        Ok(response) => response,
        Err(e) => return (RequestType::Head, Err(e)),
    };

    // Successful head request
    if response.status() != StatusCode::METHOD_NOT_ALLOWED {
        return (RequestType::Head, Ok(response));
    }

    let result = client.get(url).send().await;

    (RequestType::Get, result)
```

Reply:
Yeah agree that does feel better